cxx-qt-lib/build.rs

```diff
@@ -80,5 +81,62 @@ fn main() {
     }
 
     // Write the generated sources to a qt_types_cxx.json file
-    write_cxx_sources(generated, &format!("{}/qt_types_cxx.json", path));
+    write_cxx_sources(&generated, &format!("{}/qt_types_cxx.json", path));
+
+    // Write the generated sources to CXX_QT_LIB_OUT_DIR if set
+    println!("cargo:rerun-if-env-changed=CXX_QT_LIB_OUT_DIR");
+    if let Ok(env_var) = std::env::var("CXX_QT_LIB_OUT_DIR") {
+        let directory = PathBuf::from(env_var);
+        if !directory.is_dir() {
+            panic!(
+                "CXX_QT_LIB_OUT_DIR {} is not a directory",
+                directory.display()
+            );
+        }
+
+        let include_directory_path = PathBuf::from(format!("{}/include", &directory.display()));
+        std::fs::create_dir_all(&include_directory_path)
+            .expect("Could not create CXX_QT_LIB_OUT_DIR include dir");
+        let source_directory_path = PathBuf::from(format!("{}/src", &directory.display()));
+        std::fs::create_dir_all(&source_directory_path)
+            .expect("Could not create CXX_QT_LIB_OUT_DIR source dir");
+
+        std::fs::copy(
+            format!("{}/include/qt_types.h", dir_manifest),
+            format!("{}/qt_types.h", include_directory_path.display()),
+        )
+        .expect("Could not copy qt_types.h to CXX_QT_LIB_OUT_DIR");
+        std::fs::copy(
+            format!("{}/src/qt_types.cpp", dir_manifest),
+            format!("{}/qt_types.cpp", source_directory_path.display()),
+        )
+        .expect("Could not copy qt_types.cpp to CXX_QT_LIB_OUT_DIR");
+
+        File::create(&format!("{}/cxx.h", include_directory_path.display()))
+            .unwrap_or_else(|_| panic!("Could not create cxx.h file"))
+            .write_all(cxx_gen::HEADER.as_bytes())
+            .expect("Could not write cxx.h file");
+
+        for class in generated {
+            let include_file_path = PathBuf::from(format!(
+                "{}/{}.h",
+                include_directory_path.display(),
+                class.name
+            ));
+            File::create(&include_file_path)
```

Comment:
We have this file create, write_all, some form of failure block 4 times, DRY points to this being put in a method ?

Or also maybe the whole block needs to be split up, as the header and source paths are repeated for creating the directory, copying the qt_types, creating the individual files.  Maybe there should be one method that does each of those steps once but is passed as parameter to tell it the destination and source ?